### ISSUE TITLE
Remove usage of potentially slow endpoints

### ISSUE THREAD

**Issue by tlrobinson** [metabase/metabase#11856]
This is a parent issue to track various endpoints which can be extremely slow if an instance has a large number of saved questions and/or tables:

1. `GET /api/card?f=all`
    * problem: loads every saved question on an instance
    * where:
        * dashboards #11316
2. `GET /api/database?include_tables=true`
    * problem: loads every table on an instance
    * where:
        * admin data permissions editor
        * data reference 
3. `GET /api/database?include_tables=true&include_cards=true`
    * problem: loads every saved question and table on an instance
    * where:
        * new query page
        * query builder
            * Improve Metadata loading latency in the Query Builder Stage 1 #11805 
            * Improve Metadata loading latency in the Query Builder Stage 2 #11806
            * Rethink "Saved Questions" database: #10163
4. `GET /api/database/:id/metadata`
    * problem: loads every table and field for a database
    * where:
        * data reference
        * dashboards

See also https://github.com/metabase/metaboat/issues/69

Call trees:

```
GET /api/card?f=all
    CardApi.list() (metabase/services.js)
        fetchCards() (metabase/dashboard/dashboard.js)
            <Dashboard> (metabase/dashboard/components/Dashboard.jsx)
            <AddSeriesModal> (metabase/dashboard/components/AddSeriesModal.jsx)
            <AddToDashSelectQuestionsModal> (metabase/dashboard/components/AddToDashSelectQuestionsModal.jsx)
```

```
GET /api/database/:id/metadata
     MetabaseApi.db_metadata() (metabase/services.js)
         Databases.actions.fetchDatabaseMetadata() (metabase/entities/databases.js)
            fetchDatabaseMetadata() (metabase/redux/metadata.js)
               wrappedFetchDatabaseMetadataAndQuestion() (metabase/reference/reference.js)
                   MISC REFERENCE
               fetchSegmentFields() (metabase/redux/metadata.js)
                   wrappedFetchSegmentFields() (metabase/reference/reference.js)
                       MISC REFERENCE
               fetchRealDatabasesWithMetadata() (metabase/redux/metadata.js)
                   wrappedFetchGuide() 
                       MISC REFERENCE
               fetchDatabasesWithMetadata()  (metabase/redux/metadata.js)
                   NONE
                addCardToDashboard() (metabase/dashboard/dashboard.js)
                <AddSeriesModal> (metabase/dashboard/components/AddSeriesModal.jsx)
                <DashCardCardParameterMapper> (metabase/dashboard/containers/DashCardCardParameterMapper.jsx)
               <FieldApp> (metabase/admin/datamodel/containers/FieldApp.jsx)
```

```
GET /api/database?include_tables=true&include_cards=true
    MetabaseApi.db_list_with_tables() (metabase/services.js)
        NONE
    fetchDatabases() (metabase/redux/metadata.js)
        NONE
        fetchDatabasesWithMetadata() (metabase/redux/metadata.js)
            NONE
    Databases.actions.fetchList({ include_tables: true, include_cards: true }) (metabase/entities/databases.js)
        initializeQB() (metabase/query_builder/actions.js)
            <QueryBuilder> (metabase/query_builder/containers/QueryBuilder.jsx)
            locationChanged() (metabase/query_builder/actions.js)
                <QueryBuilder> (metabase/query_builder/containers/QueryBuilder.jsx)
    Database.loadList({ query: { include_tables: true, include_cards: true } }) (metabase/entities/databases.js)
        <NewQueryOptions> (metabase/new_query/containers/NewQueryOptions.jsx)
```

```
GET /api/database?include_tables=true
    MetabaseApi.db_real_list_with_tables() (metabase/services.js)
        NONE
    fetchRealDatabases() (metabase/redux/metadata.js)
        <DataPermissionsApp> (metabase/admin/permissions/containers/DataPermissionsApp.jsx)
        wrappedFetchDatabases (metabase/reference/reference.js)
           <DatabaseListContainer> (metabase/reference/databases/DatabaseListContainer.jsx) 
        fetchRealDatabasesWithMetadata() (metabase/redux/metadata.js)
            wrappedFetchGuide() (metabase/reference/reference.js)
                <GettingStartedGuideContainer> 
```

**Comment by tlrobinson** [comment]
Related to #10035 "Standardize how we tell API endpoints to include sub-resources + remove functionality-specific custom endpoints" 

I believe one part of the problem is functionality-specific endpoints that we have misused on the frontend out of convenience or ignorance. ["Overfetching" and "underfetching"](https://stackoverflow.com/questions/44564905/what-is-over-fetching-or-under-fetching) are common problem in REST-like APIs, and one which is solved by things like GraphQL or a way to specify specific included fields and "expand" sub-resources e.x. in the Asana API I mentioned in #10035. Of course the ability to fetch arbitrarily nested objects like that is also risky. That could be solved by adding restrictions on what you can fetch.

**Comment by salsakran** [comment]
bumping the priority of this since some of the reported slowdown times are brutal (eg minutes)

**Comment by bplmp** [comment]
Yes, I'm seeing about 30 seconds load time for the `GET /api/database?include_tables=true&include_cards=true` call, which makes it very frustrating for users. We have about 20k questions in our Metabase instance.

**Comment by Tadimsky** [comment]
Yeah, we're seeing the same issue for `/api/database?include_tables=true&include_cards=true`.

![image](https://user-images.githubusercontent.com/1149244/76107836-57210d00-5f8e-11ea-81ac-e7d160947aed.png)

It's making users wait for almost a minute to load any question.

**Comment by paulrosenzweig** [comment]
Fixed in #11805
